### ISSUE TITLE
Add bcrypt hashes support to basic_auth filter

### ISSUE THREAD

**Issue by vixns** [envoyproxy/envoy#36278]
Currently the auth_basic filter only supports SHA hashed passwords.

ExternalSecrets provide a htpasswd templating function from sprig that generates only bcrypt hashes

Supporting brcypt hashes for basic_auth will improve security and simplify the workflow I use -> `hashicorp vault>external secret>k8s secret>envoy gateway securitypolicy>envoy basic_auth filter`

https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/basic_auth_filter#configuration
https://masterminds.github.io/sprig/crypto.html
https://external-secrets.io/latest/guides/templating/

**Comment by Athishpranav2003** [comment]
I guess i can try to work on this issue
Any specific suggestions?

**Comment by Athishpranav2003** [comment]
How do we want to configure the filter to use SHA vs bcrypt hashing algo?

**Comment by Athishpranav2003** [comment]
@vixns i dont think we can use `bcrypt` hashing here since the salt is generated randomly. So hence we wont be able to verify the password by comparing the hash.

Just for the sake of completion maybe i can add MD5 support to the function along with SHA(even tho its more insecure) just to keep it same with what apache docs specify

**Comment by Athishpranav2003** [comment]
I guess i can work on this feature

**Comment by Athishpranav2003** [comment]
@vixns So openssl doesnt support bcrypt. How do u suggest to proceed with this issue?
we would either need to ship the implementation along with code here
or use some other library

**Comment by vixns** [comment]
Hi @Athishpranav2003, the OpenBSD implementation seems fine, thanks for working on this

**Comment by Athishpranav2003** [comment]
So @vixns do we indeed ship the openbsd implementation along with this?

I am not sure how does dependencies work in envoy. I can implement it but not sure how to include the dependency

**Comment by vixns** [comment]
@Athishpranav2003, did you consider using https://github.com/trusch/libbcrypt ?

About dependencies: https://github.com/envoyproxy/envoy/blob/main/DEPENDENCY_POLICY.md

**Comment by Athishpranav2003** [comment]
This also seems to solve the problem, except that I didn't know if it's the standard bcrypt implementation. If it's fine for maintainer's I will use it i guess

**Comment by Athishpranav2003** [comment]
@vixns so for external dependencies there are many more details that need to be provided and i feel that libbcrypt will not be acceptable by the maintainers. I guess we need someone from maintainer team to step in to approve one of the choices we have here

**Comment by Athishpranav2003** [comment]
@jmarantz can u help us proceed with the dependency part?
I have the code ready for the PR except that need to accomodate the external dependency

**Comment by Athishpranav2003** [comment]
@KBaichoo could you please help me with the dependency part? or redirect me to someone who is apt for this query?
I have the PR ready, only thing needed is to include the openBSD dependency

**Comment by Athishpranav2003** [comment]
@alyssawilk Idk whom to tag for this. No one came forward to help me resolve the dependency part. Could you please help me

**Comment by Athishpranav2003** [comment]
The problem i am not aware of setting up the external dependency in this case since here the dependency is in OpenBSD repo(not a library)

**Comment by alyssawilk** [comment]
my intuition is you're going to have to find a way to pull the code that isn't pulling all of OpenBSD.  I'll poke around internally to see if relevant folks know of a trusted standalone library but if you can't find one I suspect you're going to have to find a way to make an extension point and do this in your own repo rather than upstream

**Comment by Athishpranav2003** [comment]
@alyssawilk so the code part for this is relatively stable(doesn't change unless there is a CVE) so I presume if I can properly add the dependency in the losted dependency and directly pull the code part alone should be fine.

But I want a confirmation from you/any maintainer for this.

Plus how do I add the dependency for pulling the code part?

**Comment by Athishpranav2003** [comment]
@alyssawilk thanks for letting me know your opinion. 

I am going ahead with adding this repo as a external dependency https://github.com/hilch/Bcrypt.cpp

I hope this works in our case
Will make the pr tomorrow

**Comment by jmarantz** [comment]
Just to clarify; is the plan to make this an extension? I would assume wouldn't want to add this dependency in core?

**Comment by Athishpranav2003** [comment]
@jmarantz this should be an dependency for the basic auth extension 
Do you suggest any other method?

**Comment by jmarantz** [comment]
I don't know the details of how this should work, but I'm guessing that we wouldn't want enabling auth to imply enabling bcrypt; we'd want to use a separate extension point for that feature. I'll defer to Alyssa though.

**Comment by Athishpranav2003** [comment]
Technically we store hashed passwords during config and use it to verify it on run
The verification will be based on the format of hash(no enabling)

If the hash prefix is SHA then we use SHA, if its $2y$ then its bcrypt.

In this way there is no difference infact(its not affecting existing workflow)

**Comment by nikotih** [comment]
Hi guys

is bcrypt support still planned for envoy (in core or extension?). I see PR was auto closed
We would also benefit from using bcrypt instead of sha

**Comment by jmarantz** [comment]
@yanavlasov FYI

IMO this should be in 'contrib' or maybe an 'extension' if a suitable maintainer-sponser can be found, but should not be in 'core'. E.g. we probably don't want this in our build internally, and this is straightforward as an extension or contrib.

**Comment by nikotih** [comment]
I know you have your reasons to not have it in core, i am also quite distant from C++ and envoy code base.
However, nobody should use sha1 hashing for passwords. Its long considered insecure
https://httpd.apache.org/docs/2.4/misc/password_encryptions.html
https://www.schneier.com/blog/archives/2005/02/cryptanalysis_o.html
https://www.dcode.fr/sha1-hash

So i would even go beyond supporting bcrypt - i would suggest you deprecate usage of sha-1 at all and make bcrypt new default :) This way you will be promoting more secure default algorithm

**Comment by jewertow** [comment]
@jmarantz can you take a look at this PR: https://github.com/envoyproxy/envoy/pull/38666? This change aims to enable users to implement bcrypt in "contrib" and use it in the basic_auth extension.

**Comment by vixns** [comment]
up

**Comment by Pluies** [comment]
I'm coming to this via Envoy Gateway, which inherits that limitation from the Envoy Proxy, and I'd like to echo the concerns of people on this thread here about only supporting the insecure (not my words! [Straight from the Apache doc!](https://httpd.apache.org/docs/2.4/misc/password_encryptions.html)) `SHA` method.

I'd love to see Envoy support at the very least the md5 and bcrypt options, which are both secure (or at least not explicitly called out as insecure in the docs), and the most widely-used in the wild.

**Comment by Preisschild** [comment]
Not a C++ developer, but maybe [Botan](https://github.com/randombit/botan) would be an acceptable dependency

https://botan.randombit.net/doxygen/bcrypt_8h.html

**Comment by Domini** [comment]
bcrypt would be amazing, but could it be easier to support at least SHA-256 or SHA-512 in addition to the current, very insecure SHA-1?
